torchmetrics/functional/audio/pit.py

```diff
@@ -153,10 +153,17 @@ def pit(
 
     # calculate the metric matrix
     batch_size, spk_num = target.shape[0:2]
-    metric_mtx = torch.empty((batch_size, spk_num, spk_num), dtype=preds.dtype, device=target.device)
-    for t in range(spk_num):
-        for e in range(spk_num):
-            metric_mtx[:, t, e] = metric_func(preds[:, e, ...], target[:, t, ...], **kwargs)
+    metric_mtx = None
+    for tar_speech_idx in range(spk_num):  # we have spk_num speeches in target in each sample
+        for est_speech_idx in range(spk_num):  # we have spk_num speeches in preds in each sample
+            if metric_mtx is not None:
+                metric_mtx[:, tar_speech_idx, est_speech_idx] = metric_func(
+                    preds[:, est_speech_idx, ...], target[:, tar_speech_idx, ...], **kwargs
+                )
+            else:
+                first_ele = metric_func(preds[:, est_speech_idx, ...], target[:, tar_speech_idx, ...], **kwargs)
+                metric_mtx = torch.empty((batch_size, spk_num, spk_num), dtype=first_ele.dtype, device=first_ele.device)
```

Comment:
am i missing something or is it correct that you are overriding `metric_mtx` in each loop?
based on the old code it would seem like you need to append?

Reply:
Actually, `metric_mtx` is just initialized in the first loop
At a second and larger times, `metric_mtx` is not `None`, thus it won't be initialized anymore, just append the results

Reply:
can you init the variable before and use device or dtype from pred?

Reply:
@quancs got it :]
@justusschock I think that was the code did before, but the `metric_func` may output a different dtype so you need to do atleast one computation before knowing how to initialize the large tensor.